<a href="https://colab.research.google.com/github/pandiarajan-src/PyWorks/blob/master/Essentials_of_Probability_Assignment_1_2_solved.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Please make a copy of this notebook and try solving the below questions in youe notebook copy. Solutions for these will be discussed in the assignment review session on thursday post which, we will also share the solution notebook.


# **Good Luck**


# Question 1
## Coin Toss and Betting
We have given a function coin_toss which simulate a single coin toss.

The function coin_trial simulates a trail of 100 coint tosses by calling coin_toss function each time.

Note that a fair coin is suppose to give you 50 heads and 50 tails.
But that is not always the case. If you toss 100 times you might get less or more than 50 heads.


But what if you repeat the experiment 10, 100, 1000 .. 10000 times. You will see that then the average number of heads will reach close to 50.

The simulate function simulates this repeating experiment. If you call simulate with n = 1 then it is same as 100 coin tosses. If you call it with n = 10 then you repeat a trail of 100 coin tosses, 10 times.

There is a very interesting probability theory in play which you will learn in the next class.

But even if one doesn't know the law they can verify this by simulation

In [ ]:
import random
def coin_toss():
  if random.random() <= 0.5:
    return 1
  return 0

def coin_trial():
  heads = 0
  for i in range(100):
      heads +=coin_toss()
  return heads
def simulate(n):
    trials = []
    for i in range(n):
        trials.append(coin_trial())
    return(sum(trials)/n)

In [ ]:
simulate(1)

48.0

In [ ]:
simulate(10)

49.1

In [ ]:
simulate(100)

49.52

In [ ]:
simulate(1000)

49.758

In [ ]:
simulate(10000)

50.0971

Try simulating this 100 coin toss 10, 100, 1000, 10000 times and notice if the average number of heads you are getting is approaching close to 50. Can you tell intuitively why this happens?


## Betting offer
Your friends comes up with a fair coin and proposes the following. You will toss the coin untill heads comes up. For each tails you will pay your friend 10\$, But if a head comes up your friend will pay you 15\$

The sample space of such an experiment looks like this {H, TH, TTH, .... TTTTTH ....}. As soon as a head comes up you stop tossing and your friend pays you 100\$. You need to pay your friend 10\$ for each tails that came up. So to make a profit you need to get head as quickly as possible.

You can use the simulate a fair coin like we did above. Now you use power of simulation.

Write a simulation logic and a trial logic with which you can decide if you should take this bet or not.

What if your friend offered you 100\$ once head comes up? Will you take the bet?

Write a simulation function to decide.

In [2]:
# This is IK's solution
import random

def coin_toss():
  if random.random() <= 0.5:
    return 1
  return 0

def simulate(n, recieve, give):
  if n == 0:
    return recieve-0
  trials = []
  for index in range(n):
    numberofTails=0
    while coin_toss()==0:
      numberofTails=numberofTails+1

    trials.append(numberofTails)
  avgTails = sum(trials)/(n) if trials else 0
  print(trials)
  return recieve - give*avgTails

simulate(100,15,10)



[1, 2, 4, 1, 5, 1, 0, 0, 1, 1, 1, 1, 0, 1, 1, 2, 0, 1, 0, 0, 1, 1, 3, 0, 0, 3, 0, 0, 2, 0, 1, 3, 0, 0, 0, 0, 4, 1, 1, 0, 1, 2, 0, 1, 0, 0, 0, 1, 0, 1, 0, 3, 3, 1, 0, 0, 0, 0, 2, 4, 0, 2, 4, 0, 2, 0, 0, 0, 0, 1, 3, 0, 1, 1, 0, 0, 0, 4, 0, 0, 0, 0, 0, 0, 1, 0, 2, 1, 0, 2, 0, 0, 0, 2, 2, 2, 1, 0, 0, 2]


5.5

In [9]:
# This is own solution of the above same problem
import random

HEAD = 1
TAIL = 0

def toss_coin():
  if random.random() <= 0.5:
    return TAIL
  return HEAD

def simulate_tail_until_head(times, lose_4_tail, gain_4_head):
  # If you are not going to try the simulation even for one time, no use
  if times == 0:
    return 0

  trails_tail_count = []
  trails_cost = []


  for index in range(times):

    tail_count_for_iteration = 0
    while toss_coin() == HEAD:
      tail_count_for_iteration += 1

    trails_tail_count.append(tail_count_for_iteration)
    trails_cost.append(gain_4_head - (tail_count_for_iteration * lose_4_tail))

  avg_tails = sum(trails_tail_count)/len(trails_tail_count)
  avg_cost = sum(trails_cost)/len(trails_cost)

  # calculated_avg_cost = (gain_4_head - (avg_tails * lose_4_tail))

  print(trails_tail_count)
  print(avg_tails)

  print(trails_cost)
  print(avg_cost)
  # print(calculated_avg_cost)


simulate_tail_until_head(10000, 10, 15)



[1, 0, 0, 2, 3, 1, 0, 1, 0, 1, 0, 2, 3, 0, 0, 0, 0, 0, 0, 3, 2, 1, 0, 0, 0, 0, 1, 4, 0, 0, 0, 0, 1, 0, 3, 1, 0, 0, 0, 2, 2, 4, 1, 2, 0, 0, 0, 1, 0, 0, 1, 0, 0, 1, 2, 0, 0, 1, 2, 0, 1, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 1, 1, 0, 3, 1, 3, 0, 0, 0, 3, 1, 1, 2, 1, 1, 0, 1, 0, 1, 0, 0, 0, 0, 1, 2, 0, 1, 1, 0, 0, 0, 0, 0, 2, 1, 1, 0, 0, 1, 2, 3, 0, 0, 0, 2, 2, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 2, 0, 0, 1, 4, 0, 0, 2, 1, 0, 4, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 1, 1, 0, 0, 0, 1, 1, 0, 2, 0, 1, 0, 2, 1, 0, 2, 0, 0, 6, 2, 3, 0, 1, 2, 0, 0, 0, 0, 1, 0, 1, 1, 0, 1, 1, 0, 0, 0, 0, 1, 0, 1, 2, 0, 1, 2, 0, 0, 0, 0, 2, 0, 3, 0, 2, 0, 1, 1, 2, 3, 1, 1, 1, 2, 0, 0, 0, 1, 1, 0, 0, 2, 0, 0, 1, 0, 3, 0, 3, 1, 2, 1, 0, 0, 2, 0, 0, 1, 3, 3, 1, 2, 2, 2, 0, 1, 2, 0, 0, 1, 3, 0, 2, 3, 4, 1, 1, 5, 1, 0, 1, 1, 0, 0, 0, 1, 0, 2, 1, 1, 0, 2, 0, 1, 5, 0, 0, 1, 0, 2, 0, 1, 1, 0, 1, 0, 1, 0, 2, 2, 1, 0, 1, 1, 1, 0, 5, 0, 0, 0, 2, 1, 2, 0, 1, 1, 1, 3, 0, 0, 0, 2, 0, 1, 0, 1, 1, 0, 1, 1, 0, 5, 3, 0, 2, 1, 0, 0, 3, 1, 0, 0, 4, 0, 

# Question 2
## Dice Roll
Write a python function to simulate a casino where 2 dice are rolled together and the player wins if there is two six and gets 100\$. If the casino fee is 5 cents per game, what will be average profit?

In [9]:
import random

def roll_a_dice():
  """
  This is to simulate the rolling of a single dice.
  input => No
  output => produces a number between 1 to 6
  """
  return random.randint(1, 6)

def simulate_casino_dice_roll(times, price_amount, fee_per_game):
  """
  player wins if both dice get 6 and paid with price_amount
  if not, no price
  per game charged for $5
  """

  game_cost = []
  game_count_tracker = []
  for iteration in range(times):

    trials = []
    while True:
      face1 = roll_a_dice()
      face2 = roll_a_dice()
      trials.append((face1, face2))
      if face1 == 6 and face2 == 6:
        break

    game_count = len(trials)
    game_count_tracker.append(game_count)
    #print(f"{game_count} : {trials}")
    cost = price_amount - (game_count * fee_per_game)
    game_cost.append(cost)

  final_cost = sum(game_cost)/len(game_cost)
  print(game_count_tracker)
  print(game_cost)
  print(f"The final average cost of a game{final_cost}")


simulate_casino_dice_roll(100000, 100, 5)





[18, 1, 2, 54, 19, 10, 6, 15, 33, 25, 6, 10, 36, 145, 54, 19, 63, 7, 33, 10, 2, 40, 2, 80, 6, 4, 13, 35, 12, 8, 77, 49, 11, 147, 3, 29, 8, 19, 39, 20, 42, 10, 119, 51, 5, 59, 64, 6, 3, 2, 22, 9, 1, 75, 21, 7, 47, 3, 20, 28, 18, 1, 56, 2, 28, 4, 4, 27, 4, 37, 22, 14, 2, 52, 74, 21, 35, 3, 134, 91, 10, 85, 5, 18, 116, 3, 39, 112, 28, 39, 8, 10, 21, 42, 243, 22, 6, 41, 15, 4, 23, 5, 4, 77, 15, 31, 117, 51, 12, 36, 24, 76, 18, 1, 3, 19, 43, 25, 11, 6, 21, 6, 32, 13, 11, 1, 4, 15, 47, 12, 13, 4, 103, 7, 50, 37, 69, 31, 33, 4, 26, 97, 46, 29, 15, 98, 25, 328, 16, 15, 49, 14, 21, 30, 101, 30, 87, 56, 53, 11, 25, 15, 15, 36, 35, 7, 24, 4, 7, 55, 13, 2, 72, 4, 139, 10, 53, 5, 97, 33, 32, 187, 11, 6, 82, 11, 3, 27, 4, 189, 2, 14, 21, 24, 27, 22, 40, 20, 56, 19, 20, 7, 11, 17, 16, 4, 53, 19, 17, 93, 69, 13, 13, 11, 54, 97, 14, 4, 15, 2, 3, 10, 243, 73, 15, 41, 80, 31, 4, 5, 52, 16, 54, 41, 2, 64, 17, 23, 34, 17, 4, 116, 23, 18, 3, 10, 25, 79, 19, 40, 68, 3, 36, 8, 4, 5, 28, 73, 57, 2, 152, 12, 13

In [14]:
# IK solution

import random
def die_throw():
  face=random.randint(1,6)
  return face

def simulate(n,win,fee):

  trials = []
  for index in range(n):
    f1=die_throw()
    f2=die_throw()
    count = 0
    while f1!= 6 or f2!=6:

      f1=die_throw()
      f2=die_throw()
      count += 1


    totalfee = fee*count
    profit = win - totalfee
    trials.append(profit)
  return sum(trials)*1.0/n

win = 100
fee = 5
print('The profit is',simulate(100,win, fee))

The profit is -83.5


# Question 3
## Covid Test Data
### Cure Diagonistics is conducting covid test on people. They have some data on covid test done on last 1 week. You are an analyst at Cure Diagonistics.

### You are given a covid test data where 100 patients underwent RT-PCR and Fast Test. You are also given whether the patients were actually infected or not and the gender of the person. Each of the column has a binary entry.

### covid_fast_kit: 0 - Covid Negative, 1- Covid Positive


### covid_rt_pcr: 0 - Covid Negative, 1- Covid Positive


### actual_diagnostic: 0 - Covid Negative, 1- Covid Positive

### Sex - 0 : Male, 1: Female






In [ ]:
from google.colab import files
data_to_load = files.upload()

In [15]:
import pandas as pd

df= pd.read_csv("covid.csv")

FileNotFoundError: ignored

In [ ]:
df.head()

patient_id  sex  covid_fast_kit  covid_rt_pcr  actual_diagnostic
0           0    0               1             0                  0
1           1    1               0             1                  1
2           2    0               1             0                  1
3           3    0               0             0                  0
4           4    0               1             0                  1

##
### Given this data you need to figure out the following.

1. Cure Diagonistics gives out free medicine kit to covid patient who are actually infected. Given they have 100 patients coming in ever week, how many free medicine kit should they prepare for next week given the data above?
Who is more likely to be covid infected - Male or Female?

2. covid_rt_pcr takes 48 hrs to get results. Cure Diagonistics is planning to have covid_fast_kit test which reduces testing time from 48 hrs to 40 mins. Both covid_rt_pcr and covid_fast_kit test were done last week for each patient.
  

    
> i. If a patient has tested positive using covid_fast_kit, what is the probability that he or she is actually infected?

> ii. If a patient has tested positive using covid_rt_pcr, what is the probability that he or she is actually infected?

From i. and ii what will you inform the patients about the trade off?







# Question 4
## Infinte monkey Theorem
The infinite monkey theorem states that a monkey hitting keys at random on a typewriter keyboard for an infinite amount of time will almost surely type any given text, such as the complete works of William Shakespeare.

Suppose the type writer can only type between A-Z. Ignoring punctuation, spacing, and capitalization write a function that calculates the probability of writing the tiltle "Hamlet"?

The text of Hamlet contains approximately 130,000 letters. Can you comment on the chance of the monkey typing out Hamlet by randomly hitting keys of the typewriter?




# Question 5
#Spam Classification

##Exploring the Dataset

Let's start by opening the SMSSpamCollection file with the read_csv() function from the pandas package. We're going to use:

sep='\t' because the data points are tab separated

header=None because the dataset doesn't have a header row

names=['Label', 'SMS'] to name the columns

In [ ]:
!wget https://archive.ics.uci.edu/ml/machine-learning-databases/00228/smsspamcollection.zip
!unzip smsspamcollection.zip


--2023-04-01 04:24:46--  https://archive.ics.uci.edu/ml/machine-learning-databases/00228/smsspamcollection.zip
Resolving archive.ics.uci.edu (archive.ics.uci.edu)... 128.195.10.252
Connecting to archive.ics.uci.edu (archive.ics.uci.edu)|128.195.10.252|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 203415 (199K) [application/x-httpd-php]
Saving to: ‘smsspamcollection.zip’

smsspamcollection.z 100%[===================>] 198.65K  --.-KB/s    in 0.04s   

2023-04-01 04:24:46 (4.69 MB/s) - ‘smsspamcollection.zip’ saved [203415/203415]

Archive:  smsspamcollection.zip
  inflating: SMSSpamCollection       
  inflating: readme                  


In [ ]:
import pandas as pd

sms_spam = pd.read_csv('SMSSpamCollection', sep='\t',
header=None, names=['Label', 'SMS'])

print(sms_spam.shape)
sms_spam.head()

(5572, 2)


Label                                                SMS
0   ham  Go until jurong point, crazy.. Available only ...
1   ham                      Ok lar... Joking wif u oni...
2  spam  Free entry in 2 a wkly comp to win FA Cup fina...
3   ham  U dun say so early hor... U c already then say...
4   ham  Nah I don't think he goes to usf, he lives aro...

In [ ]:
sms_spam[sms_spam['Label'] == 'spam']

Label                                                SMS
2     spam  Free entry in 2 a wkly comp to win FA Cup fina...
5     spam  FreeMsg Hey there darling it's been 3 week's n...
8     spam  WINNER!! As a valued network customer you have...
9     spam  Had your mobile 11 months or more? U R entitle...
11    spam  SIX chances to win CASH! From 100 to 20,000 po...
...    ...                                                ...
5537  spam  Want explicit SEX in 30 secs? Ring 02073162414...
5540  spam  ASKED 3MOBILE IF 0870 CHATLINES INCLU IN FREE ...
5547  spam  Had your contract mobile 11 Mnths? Latest Moto...
5566  spam  REMINDER FROM O2: To get 2.50 pounds free call...
5567  spam  This is the 2nd time we have tried 2 contact u...

[747 rows x 2 columns]

In [ ]:
sms_spam[sms_spam['Label'] == 'ham']

Label                                                SMS
0      ham  Go until jurong point, crazy.. Available only ...
1      ham                      Ok lar... Joking wif u oni...
3      ham  U dun say so early hor... U c already then say...
4      ham  Nah I don't think he goes to usf, he lives aro...
6      ham  Even my brother is not like to speak with me. ...
...    ...                                                ...
5565   ham                                       Huh y lei...
5568   ham               Will ü b going to esplanade fr home?
5569   ham  Pity, * was in mood for that. So...any other s...
5570   ham  The guy did some bitching but I acted like i'd...
5571   ham                         Rofl. Its true to its name

[4825 rows x 2 columns]


Given the above data we want to predict if a given sms is **spam** or not (aslo called **ham**).
One way to think about this problem is using Bayes' Theorem. In terms of probability we want to find out the $P(Spam | SMS)$

\begin{align}
P(Spam | SMS) = \frac{P(Spam) P(SMS|Spam) } {P(Spam) P(SMS|Spam) + P(Ham) P(SMS|Ham)}
\end{align}
where $P(Spam)$ is the prior probability of an sms being spam. $P(SMS|Spam)$ is the likelihood of a sms being spam.

Note that we need to estimate $P(Spam | SMS)$ which is the posterior probability of a SMS being spam.


What is the prior probability of a message to be Spam?
What is $P(Spam)$

\begin{align}
P(Spam) = \frac{\#Spam\_SMS}{\#SMS}
\end{align}

In the next cell write code for calculating $P(Spam)$

In [ ]:
## your code here

What is $P(Ham)$?

In [ ]:
## your code here

**Training and Test Set**

In machine learning, we need to  split our dataset into a training set and a test set. We'll use 80% of the data for training and the remaining 20% for testing.

Training data is on which we need to estimate the probabilities. We will apply this estimate to decide spam or ham in test set and measure accuracy.

Can you tell why we need to split data to training and test set?

In [ ]:
# Randomize the dataset
data_randomized = sms_spam.sample(frac=1, random_state=1)

# Calculate index for split
training_test_index = round(len(data_randomized) * 0.8)

# Split into training and test sets
training_set = data_randomized[:training_test_index].reset_index(drop=True)
test_set = data_randomized[training_test_index:].reset_index(drop=True)

print(training_set.shape)
print(test_set.shape)

(4458, 2)
(1114, 2)


Next we need to calculate $P(SMS|Spam)$. This probability will tell us what is the likelihood of a SMS being spam.

This is tricky. Why?

First we need to understand what a **SMS** is made of. A **SMS** is sequence of words. We can represent a SMS as sequence of words $w_1, w_2 ... w_n$, where $w_i$ is the $i^{th}$  word.

When we want to know the likelihood of a

\begin{align}
P(SMS|Spam) = P(w_1, w_2 \ldots w_n|Spam)
\end{align}

$P(w_1, w_2 \ldots w_n|Spam) $ looks scary!

How do we calculate this?

Let's understand this a little more. Let's take two SMS and understand this:

Spam SMS - "WINNER!! As a valued network customer you have"
Ham SMS - "Even my brother is not like to speak with me..."

If we look at the words in the SMS we can tell which one is Spam and which one is Ham.

But how are we really saying this? The words in the SMS.
Notice the first SMS has words like "winner, customer, network" these are more likely to be associated with Spam messages. Where as second sms has words like "brother, speak. me" which are genuine words usually found in personal sms.

Hence the word sequence $w_1, w_2 \ldots w_n$ will be different for Spam and Ham sms and we want to estimate $P(w_1, w_2 \ldots w_n|Spam)$.

Remember probability can be estimated with counts. Let $c$ be the number of times word sequence $w_1, w_2 \ldots w_n$ is found in data and $c_{spam}$ is the number of times it is marked as spam

\begin{align}
P(w_1, w_2 \ldots w_n|Spam) = \frac{c_{spam}}{c}
\end{align}

When $w_1, w_2 \ldots w_n$ = "WINNER!! As a valued network customer you have" then we are trying to see how many times the same sms has come. Ideally this will be very rare that such exact sms comes multiple times so the probability estimates will be incorrect.

To make better estimate we are going to make an assumption. What if the words $w1, w2 \ldots w_n$ are independent?
\begin{align}
P(w_1, w_2 \ldots w_n|Spam) = P(w_1|Spam)P(w_2|Spam) \ldots P(w_n|Spam)
\end{align}

This follows from multiplicative rule of probability and independence assumption.
Now if we want to calculate the likelihood of a SMS is spam then we can calculate
\begin{align}
P(w_i|Spam) = \frac{N_{i|spam}}{N_{spam}}
\end{align}
where $N_{spam}$ is the number spam emails in the training set and $N_{i|spam}$ is the number of times the word is associated with a spam email.

So for each word we need to calculate the counts of time it appears in training set, overall and by each SMS type - Spam and Ham.

Then using Bayes' rule we can calculate the Posterior.

**Data Cleaning**

When a new message comes in, our multinomial Naive Bayes algorithm will make the classification based on the results

To calculate all these probabilities, we'll first need to perform a bit of data cleaning to bring the data into a format that allows us to easily extract all the information we need. Right now, our training and test sets have this format (the messages below are fictitious to make the example easier to understand):


In [ ]:
training_set['Label'].value_counts(normalize=True)

ham     0.86541
spam    0.13459
Name: Label, dtype: float64

In [ ]:
test_set['Label'].value_counts(normalize=True)

ham     0.868043
spam    0.131957
Name: Label, dtype: float64

In [ ]:
# Before cleaning
training_set.head(3)

Label                                            SMS
0   ham                   Yep, by the pretty sculpture
1   ham  Yes, princess. Are you going to make me moan?
2   ham                     Welp apparently he retired

**Letter Case and Punctuation**

Let's begin the data cleaning process by removing the punctuation and making all the words lowercase.

In [ ]:
# After cleaning
training_set['SMS'] = training_set['SMS'].str.replace(
   '\W', ' ') # Removes punctuation
training_set['SMS'] = training_set['SMS'].str.lower()
training_set.head(3)

<ipython-input-16-82891ec54ccc>:2: FutureWarning: The default value of regex will change from True to False in a future version.
  training_set['SMS'] = training_set['SMS'].str.replace(


Label                                            SMS
0   ham                   yep  by the pretty sculpture
1   ham  yes  princess  are you going to make me moan 
2   ham                     welp apparently he retired

**Creating the Vocabulary**

Vocabulary is the set of word in the training set. Vocabulary is like dictionary, it is collection of all possible set of words in the training set. First we need to keep track of all words and post that we can calculate counts for each.
We will store the set of all unique words in training data in a list or dictionary in python

In [ ]:

vocabulary = #your code here

It looks like there are 7,783 unique words in all the messages of our training set.

In [ ]:
#expected answer if vocabulary is your list or
len(vocabulary)

7783

**How do we count?**

Suppose we have 3 sentence "secret prize! secret prize! claim now", "coming to my secret party?" and "secret prize claim winner"

Then for each words we will keep a list of size 3 - same as the number of sentence.

Note in the below cell, for first word 'secret' occurs twice in first sentence, But we still count 1 for it, it occurs once in 2nd sentence and 3rd sentence. So it has 1 in all the indices. Word 'winner' appears only in last sentence and has an entry in the last index.

Similarly for other words, we keep the count.

We have total 7783 words and 4458 sentences. So for each word we need to keep a list of 4458 where each index will store how many times the word appears in that sentence.


In [ ]:
word_counts_per_sms = {'secret': [1,1,1],
                       'prize': [1,0,1],
                       'claim': [1,0,1],
                       'now': [1,0,1],
                       'coming': [0,1,0],
                       'to': [0,1,0],
                       'my': [0,1,0],
                       'party': [0,1,0],
                       'winner': [0,0,1]
                      }

word_counts = pd.DataFrame(word_counts_per_sms)
word_counts.head()

secret  prize  claim  now  coming  to  my  party  winner
0       1      1      1    1       0   0   0      0       0
1       1      0      0    0       1   1   1      1       0
2       1      1      1    1       0   0   0      0       1

To create the dictionary we need for our training set, we can use the code below:

We start by initializing a dictionary named word_counts_per_sms, where each key is a unique word (a string) from the vocabulary, and each value is a list of the length of the training set, where each element in that list is a 0.

The code [0] * 5 outputs [0, 0, 0, 0, 0]. So the code [0] * len(training_set['SMS']) outputs a list of the length of training_set['SMS'].

We loop over training_set['SMS'] using the enumerate() function to get both the index and the SMS message (index and sms).

Using a nested loop, we loop over sms (where sms is a list of strings, where each string represents a word in a message).

We increment word_counts_per_sms[word][index] by 1.

In [ ]:
word_counts_per_sms = {unique_word: [0] * len(training_set['SMS']) for unique_word in vocabulary}

#your code here to calculate word_counts_per_sms

Now that we have the dictionary we need, let's do the final transformations to our training set.

In [ ]:
word_counts = pd.DataFrame(word_counts_per_sms)
word_counts.head()

act  rem  mahfuuz  wer  ideas  nature  credit  thnk  proper  compensation  \
0    0    0        0    0      0       0       0     0       0             0   
1    0    0        0    0      0       0       0     0       0             0   
2    0    0        0    0      0       0       0     0       0             0   
3    0    0        0    0      0       0       0     0       0             0   
4    0    0        0    0      0       0       0     0       0             0   

   ...  olave  starving  reply  playin  weight  09065394973  boooo  loud  \
0  ...      0         0      0       0       0            0      0     0   
1  ...      0         0      0       0       0            0      0     0   
2  ...      0         0      0       0       0            0      0     0   
3  ...      0         0      0       0       0            0      0     0   
4  ...      0         0      0       0       0            0      0     0   

   seventeen  nudist  
0          0       0  
1          0       0  
2          0       0  
3          0       0  
4          0       0  

[5 rows x 7783 columns]

The Label column is missing, so we'll use the pd.concat() function to concatenate the DataFrame we just built with the DataFrame containing the training set. This way, we'll also have the Label and the SMS columns.

In [ ]:
training_set_clean = pd.concat([training_set, word_counts], axis=1)
training_set_clean.head()

Label                                                SMS  08712402972  \
0   ham                  [yep, by, the, pretty, sculpture]            0   
1   ham  [yes, princess, are, you, going, to, make, me,...            0   
2   ham                    [welp, apparently, he, retired]            0   
3   ham                                           [havent]            0   
4   ham  [i, forgot, 2, ask, ü, all, smth, there, s, a,...            0   

   outgoing  operator  slowly  knew  easily  l8  hcl  ...  450ppw  genus  \
0         0         0       0     0       0   0    0  ...       0      0   
1         0         0       0     0       0   0    0  ...       0      0   
2         0         0       0     0       0   0    0  ...       0      0   
3         0         0       0     0       0   0    0  ...       0      0   
4         0         0       0     0       0   0    0  ...       0      0   

   scammers  26th  leanne  usc  sorted  mudyadhu  soup  hardcore  
0         0     0       0    0       0         0     0         0  
1         0     0       0    0       0         0     0         0  
2         0     0       0    0       0         0     0         0  
3         0     0       0    0       0         0     0         0  
4         0     0       0    0       0         0     0         0  

[5 rows x 7785 columns]

**Calculating Constants First**

Now that we're done with cleaning the training set, we can begin coding the spam filter. The Bayes algorithm will need to answer these two probability questions to be able to classify new messages:

Also, to calculate $P(w_i|Spam)$ and $P(w_i|Ham)$ inside the formulas above, we'll need to use these equations:

$P(w_i|Spam) = \frac{N_{i|Spam} + \alpha}{N_{spam} + \alpha.N_v}$


$P(w_i|Ham) = \frac{N_{i|Ham} + \alpha}{N_{ham} + \alpha.N_v}$

Where $N_v$ is the size of the vocabulary (7785 in our case)
$N_{i|Spam}$ is number of times word $w_i$ is found in spam email and $N_{spam}$ total number of spam email.

Note that word can appear multiple time in a sms, but we will only count it as once. There is an advanced version where count is used but here we will only use if word is present or absent.

This $\alpha$ is added to account for words which are not seen in training set but might be in test set. For them both $N_{i|Spam}$ and $N_{i}$ is 0.
This is added to our previous definition of likelihood to avoid zero probability cases just because word is not seen in training




In [ ]:
# Isolating spam and ham messages first
spam_messages = # your code here
ham_messages = # your code here

# P(Spam) and P(Ham)
p_spam = # your code here (prior probability)
p_ham = # your code here (prior probability)

# N_Spam

n_spam = # your code here (total number of spam message)

# N_Ham

n_ham = # your code here (total number of ham message)

# N_Vocabulary
n_vocabulary = len(vocabulary)

# Laplace smoothing
alpha = 1

**Calculating Parameters**

Now that we have the constant terms calculated above, we can move on with calculating the parameters P(wi|Spam) and P(wi|Ham).

P(wi|Spam) and P(wi|Ham) will vary depending on the individual words. For instance, P("secret"|Spam) will have a certain probability value, while P("cousin"|Spam) or P("lovely"|Spam) will most likely have other values.

Therefore, each parameter will be a conditional probability value associated with each word in the vocabulary.

In [ ]:
# Initiate parameters
parameters_spam = {unique_word:0 for unique_word in vocabulary}
parameters_ham = {unique_word:0 for unique_word in vocabulary}

# Calculate parameters
for word in vocabulary:
   n_word_given_spam = # your code here # spam_messages already defined
   p_word_given_spam = # your code here  - use the likelihood formula 𝑃(𝑤𝑖|𝑆𝑝𝑎𝑚)
   parameters_spam[word] = p_word_given_spam

   n_word_given_ham = # your code here  # ham_messages already defined 𝑃(𝑤𝑖|𝑆𝑝𝑎𝑚)
   p_word_given_ham = # your code here  - use the likelihood formula 𝑃(𝑤𝑖|Ham)
   parameters_ham[word] = p_word_given_ham

**Classifying A New Message**

Now that we have all our parameters calculated, we can start creating the spam filter. The spam filter is understood as a function that:

Takes in as input a new message (w1, w2, ..., wn).
Calculates P(Spam|w1, w2, ..., wn) and P(Ham|w1, w2, ..., wn).
Compares the values of P(Spam|w1, w2, ..., wn) and P(Ham|w1, w2, ..., wn), and:
If P(Ham|w1, w2, ..., wn) > P(Spam|w1, w2, ..., wn), then the message is classified as ham.
If P(Ham|w1, w2, ..., wn) < P(Spam|w1, w2, ..., wn), then the message is classified as spam.
If P(Ham|w1, w2, ..., wn) = P(Spam|w1, w2, ..., wn), then the algorithm may request human help.
Note that some new messages will contain words that are not part of the vocabulary. We will simply ignore these words when we're calculating the probabilities.

Let's start by writing a first version of this function. For the classify() function below, notice that:

The input variable message needs to be a string.
We perform a bit of data cleaning on the string message:
We remove the punctuation using the re.sub() function.
We bring all letters to lower case using the str.lower() method.
We split the string at the space character and transform it into a Python list using the str.split() method.
We calculate p_spam_given_message and p_ham_given_message.
We compare p_spam_given_message with p_ham_given_message and then print a classification label.

In [ ]:
import re

def classify(message):
   '''
   message: a string
   '''

   message = re.sub('\W', ' ', message)
   message = message.lower().split()

   p_spam_given_message = p_spam
   p_ham_given_message = p_ham

   for word in message:
      if word in parameters_spam:
         p_spam_given_message *= #your code here

      if word in parameters_ham:
         p_ham_given_message *= #your code here

   print('P(Spam|message):', p_spam_given_message)
   print('P(Ham|message):', p_ham_given_message)

   if p_ham_given_message > p_spam_given_message:
      print('Label: Ham')
   elif p_ham_given_message < p_spam_given_message:
      print('Label: Spam')
   else:
      print('Equal proabilities, have a human classify this!')

We'll now test the spam filter on two new messages. One message is obviously spam, and the other is obviously ham.

In [ ]:
classify('WINNER!! This is the secret code to unlock the money: C3421.')

P(Spam|message): 1.3481290211300841e-25
P(Ham|message): 1.9368049028589875e-27
Label: Spam


In [ ]:
classify("Sounds good, Tom, then see u there")

P(Spam|message): 2.4372375665888117e-25
P(Ham|message): 3.687530435009238e-21
Label: Ham


**Measuring the Spam Filter's Accuracy**

The two results look promising, but let's see how well the filter does on our test set, which has 1,114 messages.

We'll start by writing a function that returns classification labels instead of printing them.

In [ ]:
def classify_test_set(message):
   '''
   message: a string
   '''

   message = re.sub('\W', ' ', message)
   message = message.lower().split()

   p_spam_given_message = p_spam
   p_ham_given_message = p_ham

   for word in message:
      if word in parameters_spam:
         p_spam_given_message *= parameters_spam[word]

      if word in parameters_ham:
         p_ham_given_message *= parameters_ham[word]

   if p_ham_given_message > p_spam_given_message:
      return 'ham'
   elif p_spam_given_message > p_ham_given_message:
      return 'spam'
   else:
      return 'needs human classification'

Now that we have a function that returns labels instead of printing them, we can use it to create a new column in our test set.

In [ ]:
test_set['predicted'] = test_set['SMS'].apply(classify_test_set)
test_set.head()

Label                                                SMS predicted
0   ham          Later i guess. I needa do mcat study too.       ham
1   ham             But i haf enuff space got like 4 mb...       ham
2  spam  Had your mobile 10 mths? Update to latest Oran...      spam
3   ham  All sounds good. Fingers . Makes it difficult ...       ham
4   ham  All done, all handed in. Don't know if mega sh...       ham

We can compare the predicted values with the actual values to measure how good our spam filter is with classifying new messages. To make the measurement, we'll use accuracy as a metric

In [ ]:
correct = 0
total = test_set.shape[0]

for row in test_set.iterrows():
   row = row[1]
   if row['Label'] == row['predicted']:
      correct += 1

print('Correct:', correct)
print('Incorrect:', total - correct)
print('Accuracy:', correct/total)

Correct: 1100
Incorrect: 14
Accuracy: 0.9874326750448833
